### PR TITLE
azure-identity prototype

### PR DESCRIPTION
**Pull request by chlowell**
This includes just `Async/ClientSecretCredential`, which should suffice to unblock development scenarios.

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: azure-identity/azure/identity/credentials/providers.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+if TYPE_CHECKING:
+    # pylint:disable=unused-import
+    from typing import Iterable, Optional, Tuple, Union
+    from azure.core import Configuration
+    from ._protocols import SupportsGetToken
+
+# pylint:disable=too-few-public-methods
+
+
+class ClientSecretTokenProvider:
```
Comment (schaabs) [inline azure-identity/azure/identity/credentials/providers.py]: I would really like to standardize on calling these XxxCredential across all languages. In this case ClientSecretCredential

File: azure-identity/azure/identity/credentials/providers.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        config.logging_policy = NetworkTraceLoggingPolicy(**kwargs)
+        # TODO: RetryPolicy should become more straightforwardly configurable
+        config.retry_policy = RetryPolicy(**kwargs)
+        config.retry_policy._retry_on_status_codes = [404, 429] + [
+            x for x in range(500, 600)
+        ]  # pylint:disable=protected-access
+        return config
+
+
+class EnvironmentTokenProvider(ClientSecretTokenProvider):
```
Comment (schaabs) [inline azure-identity/azure/identity/credentials/providers.py]: This class should eventually support both client secret auth, and certificate auth, so using ClientSecretTokenProvider as a base probably won't work long term.

File: azure-identity/azure/identity/credentials/providers.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        for token in tokens:
+            if all((scope in token["target"] for scope in scopes)):
+                if int(token["expires_on"]) - 300 > int(time()):
+                    return token["secret"]
+
+        # expired token or nothing in the cache => get a token from the endpoint
+        try:
+            request = HttpRequest("GET", self._IMDS_ENDPOINT)
+            # TODO: IMDS allows only one resource per request
+            request.format_parameters({"api-version": "2018-02-01", "resource": " ".join(scopes)})
```
Comment (schaabs) [inline azure-identity/azure/identity/credentials/providers.py]: >request.format_parameters({"api-version": "2018-02-01", "resource": " ".join(scopes)}) [](start = 12, length = 86)

I believe scopes will have to be modified to call MSI, as MSI currently expects a single resource not a list scopes.  Also I think that  you'll need to strip off the /.Default from the end of scope to get the resource name.

**Comment by chlowell** [comment]
@scbedd does this belong under `/sdk/core`?

**Comment by scbedd** [comment]
> @scbedd does this belong under `/sdk/core`?

I haven't heard much context on this. I will follow up with Scott in the morning to get that if you don't respond before then. 

Is this a shared scenario? If so, it goes under core/. Does it map to a service? Aka: Azure Active Directory or the like?

**Comment by chlowell** [comment]
I see the .NET equivalent is under `/sdk/identity`. I'll use that path here as well.

**Pull request merged**